File: core/publish.go
package core

import (
	"errors"
	"fmt"
	"github.com/atotto/clipboard"
	"github.com/nettee/bloom/model"
	"io/ioutil"
	"os"
	"os/exec"
	"path"
	"strings"
)

const hexoProject = "/Users/william/projects/nettee.github.io"

type GetDoc = func(article model.Article) (model.MarkdownDoc, error)
type Transfer = func(article model.Article, doc model.MarkdownDoc) (model.MarkdownDoc, error)
type Save = func(article model.Article, doc model.MarkdownDoc) error

type Publisher struct {
	getDoc    GetDoc
	transfers []Transfer
	save      Save
}

func (publisher *Publisher) publish(article model.Article) error {
	fmt.Println("Process article: ", article.Path())

	doc, err := publisher.getDoc(article)
	if err != nil {
		return err
	}

	for _, transfer := range publisher.transfers {
		doc, err = transfer(article, doc)
		if err != nil {
			return err
		}
	}

	err = publisher.save(article, doc)
	if err != nil {
		return err
	}

	return nil
}

var platformPublisher = map[string]Publisher{
	"xzl": {
		getDoc:    getDocGeneral,
		transfers: []Transfer{
			transferImageUrl,
		},
		save:      copyBody,
	},
	"wechat": {
		getDoc:    getDocGeneral,
		transfers: []Transfer{
			transferImageUrl,
		},
		save:      copyBody,
	},
	"hexo": {
		getDoc:  getDocGeneral,
		transfers: []Transfer{
			transferMathEquations,
			addReadMoreLabel,
			addHexoHeaderLines,
		},
		save: exportToHexo,
	},
	"zhihu": {
		getDoc: getDocGeneral,
		transfers: []Transfer{
			transferImageUrl,
		},
		save: saveBodyToTemp,
	},
}

func PublishArticle(article model.Article, platform string) error {
	if platform == "" {
		return errors.New("publish platform not provided")
	}
	fmt.Printf("Publish to platform %s...\n", platform)

	publisher, present := platformPublisher[platform]
	if !present {
		return errors.New(fmt.Sprintf("no publisher found for platform %s", platform))
	}
	return publisher.publish(article)
}

func getDocGeneral(article model.Article) (model.MarkdownDoc, error) {
	meta := article.Meta()
	docName := meta.Base.DocName
	if docName == "" {
		return model.MarkdownDoc{}, errors.New("docName is empty")
	}
	docFile := article.DocPath()
	// TODO debug mode
	fmt.Println("Markdown document: ", docFile)
	return model.ReadMarkdownDocFromFile(docFile)

}

func transferImageUrl(article model.Article, doc model.MarkdownDoc) (model.MarkdownDoc, error) {
	baseUrlPath := os.Getenv("BLOOM_BASE_URL_PATH")
	doc.TransferImageUrl(path.Join(baseUrlPath, article.Meta().Base.Name))
	return doc, nil
}

func addHexoHeaderLines(article model.Article, doc model.MarkdownDoc) (model.MarkdownDoc, error) {
	meta := article.Meta()
	headerLines := []string{
		"title: '" + doc.Title() + "'",
		"date: " + meta.Base.CreateTime.Format("2006-01-02 15:04:05"),
		"tags: [" + strings.Join(meta.Base.Tags, ", ") + "]",
		"---",
		"",
	}

	doc.PrependLines(headerLines)
	return doc, nil
}

// Hexo mathjax can only recognize `\newline` syntax instead of `\\`
func transferMathEquations(article model.Article, doc model.MarkdownDoc) (model.MarkdownDoc, error) {
	doc.TransferMathEquationFormat()
	return doc, nil
}

func addReadMoreLabel(article model.Article, doc model.MarkdownDoc) (model.MarkdownDoc, error) {
	meta := article.Meta()
	n := meta.Hexo.ReadMore
	if n < 15 {
		n = 15
	}

	if doc.Lines() < n {
		return doc, nil
	}

	readMoreLines := []string{
		"",
		"<!-- more -->",
		"",
	}
	doc.InsertLines(n, readMoreLines)

	return doc, nil
}

func exportToHexo(article model.Article, doc model.MarkdownDoc) error {
	meta := article.Meta()
	hexoPosts := path.Join(hexoProject, "source/_posts")
	name := meta.Base.Name
	targetFile := path.Join(hexoPosts, name+".md")

	// Write doc content to target file
	file, err := os.Create(targetFile)
	if err != nil {
		return err
	}
	_, err = file.WriteString(doc.Body())
	if err != nil {
		return err
	}
	err = file.Close()
	if err != nil {
		return err
	}
	fmt.Println("Write to file:", targetFile)

	// Copy images to target directory
	// TODO modify image relative path (but it just works well in Hexo)
	sourceDir := article.ImagePath()
	targetDir := path.Join(hexoPosts, name)
	err = os.MkdirAll(targetDir, os.ModePerm)
	if err != nil {
		return err
	}
	files, err := ioutil.ReadDir(sourceDir)
	if err != nil {
		return err
	}
	for _, file := range files {
		filePath := path.Join(sourceDir, file.Name())
		cmd := exec.Command("cp", "-r", filePath, targetDir)
		err := cmd.Run()
		if err != nil {
			return err
		}
	}

	fmt.Printf("Copy %d images to dir: %s\n", len(files), targetDir)

	return nil
}

func saveBodyToTemp(article model.Article, doc model.MarkdownDoc) error {
	meta := article.Meta()
	home, err := os.UserHomeDir()
	if err != nil {
		return err
	}
	tempFile := path.Join(home, "Desktop", meta.Base.DocName)
	fmt.Println(tempFile)

	file, err := os.Create(tempFile)
	if err != nil {
		return err
	}
	_, err = file.WriteString(doc.Body())
	if err != nil {
		return err
	}
	err = file.Close()
	if err != nil {
		return err
	}
	fmt.Println("Write to file:", tempFile)

	return nil
}

func copyBody(article model.Article, doc model.MarkdownDoc) error {
	err := clipboard.WriteAll(doc.Body())
	if err != nil {
		return err
	}
	fmt.Println("document body copied to clipboard")
	return nil
}

File: src/core/publish.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Callable

import pyperclip

from base.config import get_bloomstore
from model.article import Article
from model.markdown import MarkdownDoc

Transfer = Callable[[Article, MarkdownDoc], MarkdownDoc]
Save = Callable[[Article, MarkdownDoc], None]


class Platform(Enum):
    Xiaozhuanlan = 'xzl'
    WeChat = 'wechat'
    Hexo = 'hexo'
    Zhihu = 'zhihu'
    Default = 'default'

    @classmethod
    def _missing_(cls, value):
        return Platform.Default


@dataclass
class PublishProcess:
    transfers: List[Transfer] = field(default_factory=list)
    save: Save = field(default=lambda article, doc: None)

    def with_transfer(self, t: Transfer):
        self.transfers.append(t)

    def with_save(self, s: Save):
        self.save = s


def transfer_image_url(article: Article, doc: MarkdownDoc) -> MarkdownDoc:
    # TODO
    print('transfer_image_url')
    return doc


def transfer_math_equations(article: Article, doc: MarkdownDoc) -> MarkdownDoc:
    # TODO
    print('transfer_math_equations')
    return doc


def add_read_more_label(article: Article, doc: MarkdownDoc) -> MarkdownDoc:
    # TODO
    print('add_read_more_label')
    return doc


def add_hexo_header_lines(article: Article, doc: MarkdownDoc) -> MarkdownDoc:
    # TODO
    print('add_hexo_header_lines')


def copy_body(article: Article, doc: MarkdownDoc) -> None:
    # TODO
    print('copy_body')
    content = doc.body_string()
    pyperclip.copy(content)
    print('document body copied to clipboard')


def save_body_to_temp(article: Article, doc: MarkdownDoc) -> None:
    # TODO
    print('save_body_to_temp')


def export_to_hexo(article: Article, doc: MarkdownDoc) -> None:
    # TODO
    print('export_to_hexo')


platform_publishes = {
    Platform.Xiaozhuanlan: PublishProcess(
        transfers=[
            transfer_image_url,
        ],
        save=copy_body,
    ),
    Platform.WeChat: PublishProcess(
        transfers=[
            transfer_image_url
        ],
        save=copy_body,
    ),
    Platform.Hexo: PublishProcess(
        transfers=[
            transfer_math_equations,
            add_read_more_label,
            add_hexo_header_lines
        ],
        save=export_to_hexo,
    ),
    Platform.Zhihu: PublishProcess(
        transfers=[
            transfer_image_url,
        ],
        save=save_body_to_temp,
    ),
}


def publish(article: Article, platform: Platform):
    print('publishing', article.path)
    doc = article.read_doc()

    process = platform_publishes[platform]

    for t in process.transfers:
        doc = t(article, doc)

    process.save(article, doc)


if __name__ == '__main__':
    article_path = get_bloomstore() / 'LeetCode 例题精讲/03-从二叉树遍历到回溯算法'
    publish(Article(article_path), platform=Platform.Xiaozhuanlan)
